hamocc/mo_control_bgc.F90

```diff
@@ -60,5 +62,40 @@ MODULE mo_control_bgc
 ! Logical switches
       LOGICAL, SAVE :: do_ndep=.true.    ! apply n-deposition   (set via namelist)
       LOGICAL, SAVE :: do_rivinpt=.true. ! apply riverine input (set via namelist)
-      
+
+    contains
+
+      subroutine get_bgc_namelist(fname)
+      !-------------------------------------------------------------------------
+      ! Get filename for namelist file
+      !-------------------------------------------------------------------------
+        use mod_config, only: inst_suffix
+        use mod_xc, only: xchalt
+
+        implicit none
+
+        character(len=*), intent(out) :: fname
+        logical :: exists
+
+        if (.not. allocated(bgc_namelist)) then
+           inquire (file='ocn_in'//trim(inst_suffix), exist=exists)
+           if (exists) then
+              allocate(character(len=len('ocn_in'//trim(inst_suffix))) ::       &
+                   bgc_namelist)
```

Comment:
I haven't seen this syntax before, I would write
`allocate(bgc_namelist(len(...))`
This is how it is used other places in iHAMOCC

Reply:
Agree, I will change this.

Reply:
Turns out that the standard syntax returns an error in this case:

/cluster/projects/nn2980k/tomast/NORESM/noresm-dev/components/blom/hamocc/mo_control_bgc.F90(83): error #8232: If any object being allocated in the ALLOCATE statement has deferred type parameter, either type specification or SOURCE= or MOLD= specifiers shall appear.   [BGC_NAMELIST]

/cluster/projects/nn2980k/tomast/NORESM/noresm-dev/components/blom/hamocc/mo_control_bgc.F90(83): error #8196: An ALLOCATE shape must not be given for a scalar object.   [BGC_NAMELIST]

The reason is that bgc_namelist is a scalar character variable, not a character array. Still, I would like to use the test (.not. allocated(bgc_namelist)) to check if the variable is defined, so that's the reason for making it allocatable in the first place. It works, but it requires this strange syntax, so I'm changing it back again.

Reply:
If the main purpose is just the test, then one could also initialise `bgc_namelist=''` and use `len_trim(bgc_namelist)/=0` to test whether it has been assigned a value. I'm just not sure whether this syntax is according to a fortran standard. I believe it is not conforming to f90/95, but might be introduced later(?).

Reply:
Allocatable scalars and character length is part of the F2003 standard (see e.g. https://gcc.gnu.org/wiki/Fortran2003Status), and is supported by gcc from version 4.5 or 4.6, so I think it should be safe to use by now. As you say, we could also use the `len_trim` test for the same effect. 

Perhaps a better solution all together would be to use the `bgc_namelist` variable from `mo_control_bgc` directly, and only invoke `get_bgc_namelist` once. Then we don't need to define fixed length variables in the calling subroutines, the namelist file name could be of arbitrary length, and we would not need to use `trim(bgc_namelist)` to remove trailing whitespace.